### ISSUE TITLE
Support for async filter function

### ISSUE THREAD

**Issue by oh-fv** [sindresorhus/p-event#18]
Thank you for your work and effort on this nice little package, it is exactly what I was looking for!

However, I need the verify the event data by making an async API call. It would be great if the filter option was to support asynchronous functions.

**Comment by sindresorhus** [comment]
Yeah, that would be useful. PR welcome :)

**Comment by oh-fv** [comment]
Hey Sindre,

I mostly finished a hopefully acceptable solution for async filter functions. There are two things I would appreciate your opinion on before, though.

1. I'm using async/await, which means raising the required node version from 6 to 7.6. Are you okay with that?

2. I set out by fixing the failing tests :-) and noticed a design decision that was not discussed in #6. Have a look at this test currently in tests.js:

```
test('filter option returned with `multiArgs`', async t => {
	const emitter = new EventEmitter();

	delay(200).then(() => {
		emitter.emit('🦄', 1, 1);
		emitter.emit('🦄', 2, 2);
		emitter.emit('error', 10000, '💩');
		emitter.emit('🦄', 4, 3);
		emitter.emit('🦄', 3, 4);
	});

	t.deepEqual(await m(emitter, 'error', {
		filter: x => (x[0] > 9999) && (x[1] === '💩'),
		multiArgs: true
	}), [10000, '💩']);
});
```

It means that you are able to use the filter function to prevent the Promise from being rejected in case of a rejecting event defined in the `rejectionEvents` option, which here defaults to the `error` event. So right now, a filter takes priority over rejection events.

The docs, however, state that these events will reject the promise:
```
rejectionEvents
Type: string[]
Default: ['error']
Events that will reject the promise.
```

In my opinion, a filter should not take priority, and errors should cause a rejection. If I really want to filter errors, I could still supply `rejectionEvents: []` as an option.

It also makes async filters easier to implement :-)

What do you think?

**Comment by sindresorhus** [comment]
1. Sure. I've moved `master` to target Node.js 8 now.
2. I agree, filter should not take priority.

**Comment by sindresorhus** [comment]
@oh-fv Still interested in doing this?

**Comment by oh-fv** [comment]
Yes. Sorry for the delay. I'm using this package for a project at work that was put on hold for a while, but it seems like it will be continued very shortly. Please bear with me for a few more weeks if you can :-)

**Comment by sindresorhus** [comment]
@oh-fv Friendly bump :)
